controllers/fencer/reconciler.go

```diff
@@ -0,0 +1,184 @@
+	apiReset        chan<- struct{}
+	pollInterval    time.Duration
+	expiryInterval  time.Duration
+	cleanupInterval time.Duration
+	cache           *cache.Object
+
+	actionv1.Reconciler
+}
+
+// +kubebuilder:rbac:groups="",resources=node,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;delete
+// +kubebuilder:rbac:groups="",resources=persistentvolumeclaims,verbs=get;list;watch
+// +kubebuilder:rbac:groups="storage.k8s.io",resources=volumeattachments,verbs=get;list;watch;delete
+
+// NewReconciler returns a new Node label reconciler.
+//
+// The resyncInterval determines how often the periodic resync operation should
+// be run.
+func NewReconciler(api NodeFencer, apiReset chan<- struct{}, k8s client.Client, pollInterval time.Duration, expiryInterval time.Duration) *Reconciler {
+	return &Reconciler{
+		Client:          k8s,
+		log:             ctrl.Log,
+		api:             api,
+		apiReset:        apiReset,
+		pollInterval:    pollInterval,
+		expiryInterval:  expiryInterval,
+		cleanupInterval: 5 * expiryInterval,
+	}
+}
+
+func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, workers int) error {
+	// Create an generic event source. This is used by the Channel type source
+	// to collect the events and process with source event handler.
+	src := make(chan event.GenericEvent)
+
+	// Set scheme for conversions.
+	r.scheme = mgr.GetScheme()
+
+	// Initialize the cache.
+	r.cache = cache.New(r.expiryInterval, r.cleanupInterval)
+
+	// Create an event handler that uses the cache to make reconciliation
+	// decisions.
+	eventHandler := handler.NewEnqueueRequestFromCache(r.cache)
+
+	// Don't allow an overly-agressive interval to break the backend.
+	if r.pollInterval < minPollInterval {
+		r.log.Info("resetting poll interval to minimum", "interval", minPollInterval)
+	}
+
+	// Periodically populate the cache by sending StorageOS API nodes to the
+	// event source.
+	go r.pollNodes(ctx, src)
+
+	// Create an index on the Pod's node name.
+	if err := mgr.GetFieldIndexer().IndexField(ctx, &corev1.Pod{}, "spec.nodeName", func(rawObj client.Object) []string {
+		pod := rawObj.(*corev1.Pod)
+		return []string{pod.Spec.NodeName}
+	}); err != nil {
+		return err
+	}
+
+	// Create an index on the VolumeAttachment's node name.
+	if err := mgr.GetFieldIndexer().IndexField(ctx, &storagev1.VolumeAttachment{}, "spec.nodeName", func(rawObj client.Object) []string {
+		va := rawObj.(*storagev1.VolumeAttachment)
+		return []string{va.Spec.NodeName}
+	}); err != nil {
+		return err
+	}
+
+	// Create a new fencing controller.
+	c, err := NewController(r.Client, r.cache, mgr.GetScheme(), r.api, r.log)
+	if err != nil {
+		return err
+	}
+
+	// Initialize the reconciler with the fencing controller.
+	r.Reconciler.Init(mgr, c,
+		actionv1.WithName("pod-fencer"),
+		actionv1.WithScheme(mgr.GetScheme()),
+		actionv1.WithActionTimeout(25*time.Second),
+		actionv1.WithActionRetryPeriod(5*time.Second),
```

Comment:
Maybe these need to be made configurable?

Reply:
Yes, we should - thanks.  The default values look good.